File: base/common/python/pki/account.py
# Authors:
#     Endi S. Dewata <edewata@redhat.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the Lesser GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
#  along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Copyright (C) 2013 Red Hat, Inc.
# All rights reserved.
#

import inspect
import json
import logging

import pki

logger = logging.getLogger(__name__)


class AccountClient:
    """
    Class used to associate an authentication session variable with a
    connection.

    To use this class:
       * set the authentication credentials with the connection,
       * create an AccountClient and then call login().
       * further operations in this session will use the same authentication
         credentials without re-authentication.
       * call logout() to invalidate the session.
    """

    def __init__(self, parent, subsystem=None):
        """
        Creates an AccountClient for the connection.

        :param parent: PKIClient object
        :type parent: pki.client.PKIClient
        :returns: AccountClient
        """

        if isinstance(parent, pki.client.PKIConnection):

            logger.warning(
                '%s:%s: The PKIConnection parameter in AccountClient.__init__() '
                'has been deprecated. Provide SubsystemClient instead.',
                inspect.stack()[1].filename, inspect.stack()[1].lineno)

            self.subsystem_client = None
            self.pki_client = None
            self.connection = parent

            # in legacy code the subsystem name is specified in AccountClient
            # in PKIConnection
            if subsystem:
                self.subsystem_name = subsystem
            elif self.connection.subsystem:
                self.subsystem_name = self.connection.subsystem
            else:
                raise Exception('Missing subsystem for AccountClient')

        else:
            self.subsystem_client = parent
            self.pki_client = self.subsystem_client.parent
            self.connection = self.pki_client.connection

            # in newer code the subsystem name is specified in subsystem client
            # (e.g. CAClient, KRAClient)
            self.subsystem_name = self.subsystem_client.name

    @pki.handle_exceptions()
    def login(self):
        """
        Login to account REST interface.  If login is successful,
        an authentication session variable is associated with the connection.

        :returns: None
        """

        if self.pki_client:
            api_path = self.pki_client.get_api_path()
        else:
            api_path = 'v2'

        path = '/%s/account/login' % api_path

        # in legacy code the PKIConnection object might already have the subsystem name
        # in newer code the subsystem name needs to be included in the path
        if not self.connection.subsystem:
            path = '/' + self.subsystem_name + path

        response = self.connection.get(path)

        json_response = response.json()
        logger.debug('Response:\n%s', json.dumps(json_response, indent=4))

        # TODO: return Account object instead of JSON/XML
        return json_response

    @pki.handle_exceptions()
    def logout(self):
        """
        Logs out of the session.  Authentication session variables are
        invalidated for the connection

        :returns: None
        """

        if self.pki_client:
            api_path = self.pki_client.get_api_path()
        else:
            api_path = 'v2'

        path = '/%s/account/logout' % api_path

        # in legacy code the PKIConnection object might already have the subsystem name
        # in newer code the subsystem name needs to be included in the path
        if not self.connection.subsystem:
            path = '/' + self.subsystem_name + path

        self.connection.get(path)

File: base/common/src/main/java/com/netscape/certsrv/account/Account.java
// --- BEGIN COPYRIGHT BLOCK ---
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
//
// (C) 2014 Red Hat, Inc.
// All rights reserved.
// --- END COPYRIGHT BLOCK ---

package com.netscape.certsrv.account;

import java.io.StringReader;
import java.io.StringWriter;
import java.util.Collection;
import java.util.TreeSet;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.netscape.certsrv.base.RESTMessage;

/**
 * @author Endi S. Dewata
 */
@JsonInclude(Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown=true)
public class Account extends RESTMessage {

    String id;
    @JsonProperty("FullName")
    String fullName;
    @JsonProperty("Email")
    String email;
    @JsonProperty("Roles")
    Collection<String> roles = new TreeSet<>();

    public String getID() {
        return id;
    }

    public void setID(String id) {
        this.id = id;
    }

    public String getFullName() {
        return fullName;
    }

    public void setFullName(String fullName) {
        this.fullName = fullName;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public Collection<String> getRoles() {
        return roles;
    }

    public void setRoles(Collection<String> roles) {
        this.roles.clear();
        this.roles.addAll(roles);
    }

    public void addRole(String role) {
        roles.add(role);
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = super.hashCode();
        result = prime * result + ((email == null) ? 0 : email.hashCode());
        result = prime * result + ((fullName == null) ? 0 : fullName.hashCode());
        result = prime * result + ((id == null) ? 0 : id.hashCode());
        result = prime * result + ((roles == null) ? 0 : roles.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (!super.equals(obj))
            return false;
        if (getClass() != obj.getClass())
            return false;
        Account other = (Account) obj;
        if (email == null) {
            if (other.email != null)
                return false;
        } else if (!email.equals(other.email))
            return false;
        if (fullName == null) {
            if (other.fullName != null)
                return false;
        } else if (!fullName.equals(other.fullName))
            return false;
        if (id == null) {
            if (other.id != null)
                return false;
        } else if (!id.equals(other.id))
            return false;
        if (roles == null) {
            if (other.roles != null)
                return false;
        } else if (!roles.equals(other.roles))
            return false;
        return true;
    }

    @Override
    public Element toDOM(Document document) {

        Element accountElement = document.createElement("Account");
        accountElement.setAttribute("id", id);

        toDOM(document, accountElement);

        if (fullName != null) {
            Element fullNameElement = document.createElement("FullName");
            fullNameElement.appendChild(document.createTextNode(fullName));
            accountElement.appendChild(fullNameElement);
        }

        if (email != null) {
            Element emailElement = document.createElement("Email");
            emailElement.appendChild(document.createTextNode(email));
            accountElement.appendChild(emailElement);
        }

        if (!roles.isEmpty()) {
            Element rolesElement = document.createElement("Roles");
            accountElement.appendChild(rolesElement);

            for (String role : roles) {
                Element roleElement = document.createElement("Role");
                roleElement.appendChild(document.createTextNode(role));
                rolesElement.appendChild(roleElement);
            }
        }

        return accountElement;
    }

    public static Account fromDOM(Element accountElement) {

        Account account = new Account();

        String id = accountElement.getAttribute("id");
        account.setID(id);

        fromDOM(accountElement, account);

        NodeList fullNameList = accountElement.getElementsByTagName("FullName");
        if (fullNameList.getLength() > 0) {
            String value = fullNameList.item(0).getTextContent();
            account.setFullName(value);
        }

        NodeList emailList = accountElement.getElementsByTagName("Email");
        if (emailList.getLength() > 0) {
            String email = emailList.item(0).getTextContent();
            account.setEmail(email);
        }

        NodeList roleList = accountElement.getElementsByTagName("Role");
        int length = roleList.getLength();
        if (length > 0) {
            for (int i=0; i<length; i++) {
               String role = roleList.item(i).getTextContent();
               account.addRole(role);
            }
        }

        return account;
    }

    @Override
    public String toXML() throws Exception {

        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        Document document = builder.newDocument();

        Element accountElement = toDOM(document);
        document.appendChild(accountElement);

        TransformerFactory transformerFactory = TransformerFactory.newInstance();
        Transformer transformer = transformerFactory.newTransformer();
        transformer.setOutputProperty(OutputKeys.INDENT, "yes");
        transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4");

        DOMSource domSource = new DOMSource(document);
        StringWriter sw = new StringWriter();
        StreamResult streamResult = new StreamResult(sw);
        transformer.transform(domSource, streamResult);

        return sw.toString();
    }

    public static Account fromXML(String xml) throws Exception {

        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        Document document = builder.parse(new InputSource(new StringReader(xml)));

        Element accountElement = document.getDocumentElement();
        return fromDOM(accountElement);
    }

    @Override
    public String toString() {
        try {
            return toJSON();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}
